csrc/expr_simplifier.cpp

```diff
@@ -2490,6 +2490,104 @@ Val* factorizeGcd(Val* value, const Context& context) {
   return value;
 }
 
+Val* cancelTermsInPredicate(Val* value, const Context& context) {
+  auto bop = dynamic_cast<BinaryOp*>(value->definition());
+  if (!bop) {
+    return value;
+  }
+  auto op = bop->getBinaryOpType();
+  if (op != BinaryOpType::Eq && op != BinaryOpType::GE &&
+      op != BinaryOpType::GT && op != BinaryOpType::LE &&
+      op != BinaryOpType::LT && op != BinaryOpType::NE) {
+    return value;
+  }
+
+  auto get_terms = [](Val* operand) -> std::vector<Val*> {
+    if (auto add_op = dynamic_cast<BinaryOp*>(operand->definition());
+        add_op != nullptr && add_op->getBinaryOpType() == BinaryOpType::Add) {
+      return add_op->inputs();
+    } else if (auto sub_op = dynamic_cast<BinaryOp*>(operand->definition());
+               sub_op != nullptr &&
+               sub_op->getBinaryOpType() == BinaryOpType::Sub) {
+      auto inputs = sub_op->inputs();
+      inputs.at(1) = IrBuilder::negExpr(inputs.at(1));
+      return inputs;
+    } else if (auto flattened = toFlattenedAdd(operand->definition())) {
+      return flattened->inputs();
+    } else if (auto flattened = assoc_comm::flatten(operand);
+               flattened->isA<FOp>() &&
+               flattened->as<FOp>()->getOpType() == BinaryOpType::Add) {
+      return flattened->as<FOp>()->inputs();
+    } else {
+      return {operand};
+    }
+  };
+
+  const auto& lhs_terms = get_terms(bop->lhs());
+  const auto& rhs_terms = get_terms(bop->rhs());
+
+  std::unordered_set<Val*> common_lhs_terms;
+  std::unordered_set<Val*> common_rhs_terms;
+  for (auto lhs_term : lhs_terms) {
+    for (auto rhs_term : rhs_terms) {
+      // Make sure no multiple LHS terms are removed for the same RHS term
+      if (common_rhs_terms.count(rhs_term)) {
+        continue;
+      }
+      if (lhs_term->sameAs(rhs_term)) {
+        common_lhs_terms.insert(lhs_term);
+        common_rhs_terms.insert(rhs_term);
+        break;
+      }
+    }
+  }
```

Comment:
Should we make `common_rhs_terms` a `std::unordered_set<int64_t>` which keep track of the index in rhs terms, instead of `Val *`? Because if we have `i1 + i1 + i1 == i1 + i1 + i1 + i1`, the each run of this pass can only eliminate one `i1` here if storing `Val*`. But if we store indices, then we can eliminate all of the three `i1`s in one round.

Reply:
I was assuming each of `i1` should have a distinctive pointer, but that may not be always the case. Will update as you suggested.